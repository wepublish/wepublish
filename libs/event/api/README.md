# event-api

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test event-api` to execute the unit tests via [Jest](https://jestjs.io).
