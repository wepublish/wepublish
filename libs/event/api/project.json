{
  "name": "event-api",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/event/api/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/event/api/jest.config.ts"
      }
    }
  }
}
