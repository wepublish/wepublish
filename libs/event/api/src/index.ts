export * from './lib/event.module'
export * from './lib/event.model'
export * from './lib/event.service'
export * from './lib/event-dataloader.service'
