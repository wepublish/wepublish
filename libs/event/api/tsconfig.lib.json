{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/out-tsc",
    "declaration": true,
    "types": ["node"],
    "target": "es2021",
    "strictNullChecks": true,
    "noImplicitAny": true,
    "strictBindCallApply": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true
  },
  "include": ["src/**/*.ts", "../../utils/api/src/lib/dataloader/prime-dataloaders.decorator.ts"],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
