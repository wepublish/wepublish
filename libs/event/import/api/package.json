{
  "name": "@wepublish/event/import/api"
}
