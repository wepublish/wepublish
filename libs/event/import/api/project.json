{
  "name": "events-import-api",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/event/import/api/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/event/import/api/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/event/import/api/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
