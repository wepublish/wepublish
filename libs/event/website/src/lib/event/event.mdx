import {Meta, Story, Canvas, CodeOrSourceMdx} from '@storybook/blocks'
import * as stories from './event.stories'

<Meta of={stories} />

# Event

To replace the `Event` component with a custom component, one can override it like this:

<CodeOrSourceMdx>
```typescript
import {WebsiteBuilderProvider, BuilderEventProps} from '@wepublish/website'

const MyEvent = (props: BuilderEventProps) => <div>My Event</div>

export default function () {
  return (
    <WebsiteBuilderProvider Event={MyEvent}>
      <EventContainer id="123-123-123" />
    </WebsiteBuilderProvider>
  )
}
```
</CodeOrSourceMdx>

<Canvas>
  <Story of={stories.Default} />
</Canvas>
