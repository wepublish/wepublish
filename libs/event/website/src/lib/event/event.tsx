import {Theme, css, styled, useTheme} from '@mui/material'
import {BuilderEventProps, useWebsiteBuilder} from '@wepublish/website/builder'
import {MdDateRange, MdLocationCity} from 'react-icons/md'
import {EventSEO} from './event-seo'

export const EventWrapper = styled('article')`
  display: grid;
  gap: ${({theme}) => theme.spacing(1)};
`

const eventImage = (theme: Theme) => css`
  aspect-ratio: 16 / 9;
  object-fit: cover;
`

export const EventMeta = styled('div')`
  display: grid;
  grid-template-columns: auto auto;
  justify-content: space-between;
  align-items: center;
  gap: ${({theme}) => theme.spacing(2)};
`

const MetaWrapper = styled('div')`
  display: grid;
  grid-template-columns: auto auto auto;
  gap: ${({theme}) => theme.spacing(2)};
  justify-content: flex-start;
  align-items: center;
`

const EventStart = styled('div')`
  display: grid;
  grid-template-columns: auto auto;
  gap: ${({theme}) => theme.spacing(1)};
  align-items: center;
`

export const Event = ({data, loading, error, className}: BuilderEventProps) => {
  const theme = useTheme()

  const {
    elements: {Image},
    blocks: {RichText, Title},
    date
  } = useWebsiteBuilder()

  return (
    <EventWrapper className={className}>
      {data?.event && <EventSEO event={data.event} />}
      {data?.event.image && <Image css={eventImage(theme)} image={data.event.image} />}

      {data?.event && (
        <EventMeta>
          <MetaWrapper>
            <EventStart>
              <MdDateRange />
              {date.format(new Date(data.event.startsAt))}
            </EventStart>

            {data.event.endsAt && (
              <>
                <span>&ndash;</span> {date.format(new Date(data.event.endsAt))}
              </>
            )}
          </MetaWrapper>

          <MetaWrapper>
            <MdLocationCity /> {data.event.location ?? 'Veranstaltungsort noch nicht bekannt'}
          </MetaWrapper>
        </EventMeta>
      )}

      <Title title={data?.event.name} />

      <RichText richText={data?.event.description ?? []} />
    </EventWrapper>
  )
}
