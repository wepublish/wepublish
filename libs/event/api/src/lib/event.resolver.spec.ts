import {ApolloDriver, ApolloDriverConfig} from '@nestjs/apollo'
import {INestApplication} from '@nestjs/common'
import {GraphQLModule} from '@nestjs/graphql'
import {Test, TestingModule} from '@nestjs/testing'
import {Event, EventStatus, PrismaClient} from '@prisma/client'
import {ImageDataloaderService} from '@wepublish/image/api'
import {SortOrder} from '@wepublish/utils/api'
import request from 'supertest'
import {EventDataloaderService} from './event-dataloader.service'
import {CreateEventInput, EventListArgs, EventSort, UpdateEventInput} from './event.model'
import {EventResolver} from './event.resolver'
import {EventService} from './event.service'

const mockEvent = {
  id: '1234',
  createdAt: new Date('2023-01-01'),
  description: [],
  startsAt: new Date('2023-01-01'),
  endsAt: new Date('2023-01-01'),
  imageId: '123',
  location: 'Foobar',
  name: 'Foo',
  status: EventStatus.Scheduled,
  externalSourceId: null,
  externalSourceName: null,
  modifiedAt: new Date('2023-01-01')
} as Event

const eventQuery = `
  query Event($id: String!) {
    event(id: $id) {
      id
      name
      imageId
      image {
        id
        filename
      }
    }
  }
`

const eventListQuery = `
  query EventList(
    $filter: EventFilter
    $cursorId: String
    $take: Int
    $skip: Int
    $order: SortOrder
    $sort: EventSort
  ) {
    events(
      filter: $filter,
      cursorId: $cursorId,
      take: $take,
      skip: $skip,
      order: $order,
      sort: $sort
    ) {
      nodes {
        id
        name
        imageId
        image {
          id
          filename
        }
      }

      pageInfo {
        startCursor
        endCursor
        hasNextPage
        hasPreviousPage
      }

      totalCount
    }
  }
`

const createEventQuery = `
  mutation CreateEvent(
    $name: String!
    $description: RichText
    $location: String
    $startsAt: DateTime!
    $endsAt: DateTime
    $imageId: String
    $tagIds: [String!],
  ) {
    createEvent(
      name: $name
      description: $description
      location: $location
      startsAt: $startsAt
      endsAt: $endsAt
      imageId: $imageId
      tagIds: $tagIds
    ) {
      id
      name
      imageId
      image {
        id
        filename
      }
    }
  }
`

const updateEventQuery = `
  mutation UpdateEvent(
    $id: String!,
    $name: String
    $description: RichText
    $location: String
    $startsAt: DateTime
    $endsAt: DateTime
    $imageId: String
    $tagIds: [String!],
  ) {
    updateEvent(
      id: $id
      name: $name
      description: $description
      location: $location
      startsAt: $startsAt
      endsAt: $endsAt
      imageId: $imageId
      tagIds: $tagIds
    ) {
      id
      name
      imageId
      image {
        id
        filename
      }
    }
  }
`

const deleteEventQuery = `
  mutation DeleteEvent($id: String!) {
    deleteEvent(id: $id) {
      id
      name
      imageId
      image {
        id
        filename
      }
    }
  }
`

describe('EventService', () => {
  let app: INestApplication
  let eventServiceMock: {[method in keyof EventService]?: jest.Mock}
  let eventDataloaderServiceMock: {[method in keyof EventDataloaderService]?: jest.Mock}
  let imageDataloaderServiceMock: {[method in keyof EventDataloaderService]?: jest.Mock}

  beforeEach(async () => {
    eventServiceMock = {
      getEvents: jest.fn(),
      createEvent: jest.fn(),
      deleteEvent: jest.fn(),
      updateEvent: jest.fn()
    }

    eventDataloaderServiceMock = {
      load: jest.fn()
    }

    imageDataloaderServiceMock = {
      load: jest.fn()
    }

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        GraphQLModule.forRoot<ApolloDriverConfig>({
          driver: ApolloDriver,
          autoSchemaFile: true,
          path: '/',
          cache: 'bounded'
        })
      ],
      providers: [
        EventResolver,
        {
          provide: EventService,
          useValue: eventServiceMock
        },
        {
          provide: EventDataloaderService,
          useValue: eventDataloaderServiceMock
        },
        {
          provide: ImageDataloaderService,
          useValue: imageDataloaderServiceMock
        },
        {
          provide: PrismaClient,
          useValue: jest.fn() // not used due to mocks but needs to be provided
        }
      ]
    }).compile()

    app = module.createNestApplication()
    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  test('event', async () => {
    eventDataloaderServiceMock.load?.mockResolvedValue({
      id: '1234',
      name: 'name',
      imageId: '123'
    })

    imageDataloaderServiceMock.load?.mockResolvedValue({
      id: '123',
      filename: '123.webp'
    })

    await request(app.getHttpServer())
      .post('')
      .send({
        query: eventQuery,
        variables: {
          id: '1234'
        }
      })
      .expect(200)
      .expect(res => {
        expect(imageDataloaderServiceMock.load?.mock.calls[0]).toMatchSnapshot()
        expect(eventDataloaderServiceMock.load?.mock.calls[0]).toMatchSnapshot()
        expect(res.body.data.event).toMatchSnapshot()
      })
  })

  test('events', async () => {
    eventServiceMock.getEvents?.mockResolvedValue({
      nodes: [{...mockEvent, imageId: null}],
      pageInfo: {
        hasNextPage: true,
        hasPreviousPage: false,
        endCursor: '1234',
        startCursor: '123'
      },
      totalCount: 100
    })

    await request(app.getHttpServer())
      .post('')
      .send({
        query: eventListQuery,
        variables: {
          filter: {
            from: new Date('2023-01-01'),
            to: new Date('2023-01-01'),
            location: 'Foobar',
            name: 'Foo',
            tags: ['123'],
            upcomingOnly: true
          },
          order: SortOrder.Ascending,
          skip: 1,
          sort: EventSort.EndsAt,
          take: 5
        } as EventListArgs
      })
      .expect(res => {
        expect(eventServiceMock.getEvents?.mock.calls[0]).toMatchSnapshot()
        expect(res.body.data.events).toMatchSnapshot()
      })
      .expect(200)
  })

  test('create', async () => {
    eventServiceMock.createEvent?.mockResolvedValue(mockEvent)

    imageDataloaderServiceMock.load?.mockResolvedValue({
      id: '123',
      filename: '123.webp'
    })

    await request(app.getHttpServer())
      .post('')
      .send({
        query: createEventQuery,
        variables: {
          description: [],
          startsAt: new Date('2023-01-01'),
          endsAt: new Date('2023-01-01'),
          imageId: '123',
          location: 'Foobar',
          name: 'Foo',
          status: EventStatus.Scheduled,
          tagIds: ['1234']
        } as CreateEventInput
      })
      .expect(res => {
        expect(eventServiceMock.createEvent?.mock.calls[0]).toMatchSnapshot()
        expect(res.body.data.createEvent).toMatchSnapshot()
      })
      .expect(200)
  })

  test('update', async () => {
    eventServiceMock.updateEvent?.mockResolvedValue({
      ...mockEvent,
      name: 'Bar'
    })

    imageDataloaderServiceMock.load?.mockResolvedValue({
      id: '123',
      filename: '123.webp'
    })

    await request(app.getHttpServer())
      .post('')
      .send({
        query: updateEventQuery,
        variables: {
          id: mockEvent.id,
          name: 'Bar'
        } as UpdateEventInput
      })
      .expect(res => {
        expect(eventServiceMock.updateEvent?.mock.calls[0]).toMatchSnapshot()
        expect(res.body.data.updateEvent).toMatchSnapshot()
      })
      .expect(200)
  })

  test('delete', async () => {
    eventServiceMock.deleteEvent?.mockResolvedValue(mockEvent)

    imageDataloaderServiceMock.load?.mockResolvedValue({
      id: '123',
      filename: '123.webp'
    })

    await request(app.getHttpServer())
      .post('')
      .send({
        query: deleteEventQuery,
        variables: {
          id: '1234'
        }
      })
      .expect(res => {
        expect(eventServiceMock.deleteEvent?.mock.calls[0]).toMatchSnapshot()
        expect(res.body.data.deleteEvent).toMatchSnapshot()
      })
      .expect(200)
  })
})
