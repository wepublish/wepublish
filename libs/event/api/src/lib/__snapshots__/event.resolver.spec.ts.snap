// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventService create 1`] = `
[
  {
    "description": [],
    "endsAt": 2023-01-01T00:00:00.000Z,
    "imageId": "123",
    "location": "Foobar",
    "name": "Foo",
    "startsAt": 2023-01-01T00:00:00.000Z,
    "tagIds": [
      "1234",
    ],
  },
]
`;

exports[`EventService create 2`] = `
{
  "id": "1234",
  "image": {
    "filename": "123.webp",
    "id": "123",
  },
  "imageId": "123",
  "name": "Foo",
}
`;

exports[`EventService delete 1`] = `
[
  "1234",
]
`;

exports[`EventService delete 2`] = `
{
  "id": "1234",
  "image": {
    "filename": "123.webp",
    "id": "123",
  },
  "imageId": "123",
  "name": "Foo",
}
`;

exports[`EventService event 1`] = `
[
  "123",
]
`;

exports[`EventService event 2`] = `
[
  "1234",
]
`;

exports[`EventService event 3`] = `
{
  "id": "1234",
  "image": {
    "filename": "123.webp",
    "id": "123",
  },
  "imageId": "123",
  "name": "name",
}
`;

exports[`EventService events 1`] = `
[
  {
    "filter": {
      "from": 2023-01-01T00:00:00.000Z,
      "location": "Foobar",
      "name": "Foo",
      "tags": [
        "123",
      ],
      "to": 2023-01-01T00:00:00.000Z,
      "upcomingOnly": true,
    },
    "order": "Ascending",
    "skip": 1,
    "sort": "EndsAt",
    "take": 5,
  },
]
`;

exports[`EventService events 2`] = `
{
  "nodes": [
    {
      "id": "1234",
      "image": null,
      "imageId": null,
      "name": "Foo",
    },
  ],
  "pageInfo": {
    "endCursor": "1234",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "123",
  },
  "totalCount": 100,
}
`;

exports[`EventService update 1`] = `
[
  {
    "id": "1234",
    "name": "Bar",
  },
]
`;

exports[`EventService update 2`] = `
{
  "id": "1234",
  "image": {
    "filename": "123.webp",
    "id": "123",
  },
  "imageId": "123",
  "name": "Bar",
}
`;
